File: packages/migration-cli/templates/skills/business-logic-mapping/examples/controller-example.cs
// =============================================================================
// .NET Framework to .NET 10 Controller Conversion Example
// =============================================================================

// -----------------------------------------------------------------------------
// BEFORE: .NET Framework MVC Controller (Controllers/ProductController.cs)
// -----------------------------------------------------------------------------
/*
using System.Web.Mvc;
using ProjectName.Services;
using ProjectName.Models;

namespace ProjectName.Controllers
{
    public class ProductController : Controller
    {
        private readonly ProductService _productService;

        public ProductController()
        {
            _productService = new ProductService(); // Manual instantiation
        }

        public ActionResult Index()
        {
            var products = _productService.GetAllProducts();
            return View(products);
        }

        public ActionResult Details(int id)
        {
            var product = _productService.GetProductById(id);
            
            if (product == null)
            {
                return HttpNotFound();
            }
            
            return View(product);
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Create(ProductViewModel model)
        {
            if (!ModelState.IsValid)
            {
                return View(model);
            }

            var product = _productService.CreateProduct(model);
            
            TempData["Success"] = "Product created successfully.";
            return RedirectToAction("Details", new { id = product.Id });
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Edit(int id, ProductViewModel model)
        {
            if (!ModelState.IsValid)
            {
                return View(model);
            }

            _productService.UpdateProduct(id, model);
            
            TempData["Success"] = "Product updated successfully.";
            return RedirectToAction("Details", new { id });
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Delete(int id)
        {
            _productService.DeleteProduct(id);
            
            TempData["Success"] = "Product deleted successfully.";
            return RedirectToAction("Index");
        }
    }
}
*/

// -----------------------------------------------------------------------------
// AFTER: .NET 10 Controller (Controllers/ProductController.cs)
// -----------------------------------------------------------------------------

using Microsoft.AspNetCore.Mvc;
using ProjectName.Models.DTOs;
using ProjectName.Services.Interfaces;

namespace ProjectName.Controllers;

public class ProductController : Controller
{
    private readonly IProductService _productService;
    private readonly ILogger<ProductController> _logger;

    // Constructor injection (replaces manual instantiation)
    public ProductController(
        IProductService productService,
        ILogger<ProductController> logger)
    {
        _productService = productService;
        _logger = logger;
    }

    // GET: /products
    public async Task<IActionResult> Index()
    {
        var products = await _productService.GetAllProductsAsync();
        return View(products);
    }

    // GET: /products/{id}
    public async Task<IActionResult> Details(int id)
    {
        var product = await _productService.GetProductByIdAsync(id);
        
        if (product is null)
        {
            return NotFound(); // Replaces HttpNotFound()
        }
        
        return View(product);
    }

    // POST: /products
    [HttpPost]
    [ValidateAntiForgeryToken]
    public async Task<IActionResult> Create([FromForm] CreateProductDto dto)
    {
        if (!ModelState.IsValid)
        {
            return View(dto);
        }

        var product = await _productService.CreateProductAsync(dto);
        
        TempData["Success"] = "Product created successfully.";
        return RedirectToAction(nameof(Details), new { id = product.Id });
    }

    // POST: /products/{id}/edit
    [HttpPost]
    [ValidateAntiForgeryToken]
    public async Task<IActionResult> Edit(int id, [FromForm] UpdateProductDto dto)
    {
        if (!ModelState.IsValid)
        {
            return View(dto);
        }

        await _productService.UpdateProductAsync(id, dto);
        
        TempData["Success"] = "Product updated successfully.";
        return RedirectToAction(nameof(Details), new { id });
    }

    // POST: /products/{id}/delete
    [HttpPost]
    [ValidateAntiForgeryToken]
    public async Task<IActionResult> Delete(int id)
    {
        await _productService.DeleteProductAsync(id);
        
        TempData["Success"] = "Product deleted successfully.";
        return RedirectToAction(nameof(Index));
    }
}

// -----------------------------------------------------------------------------
// DTO for Create (Models/DTOs/CreateProductDto.cs)
// -----------------------------------------------------------------------------

namespace ProjectName.Models.DTOs;

public class CreateProductDto
{
    [Required]
    [MaxLength(255)]
    public string Name { get; set; } = string.Empty;

    [Required]
    [Range(0, double.MaxValue)]
    public decimal Price { get; set; }

    public string? Description { get; set; }
}

// -----------------------------------------------------------------------------
// KEY CONVERSION NOTES:
// -----------------------------------------------------------------------------
// 1. System.Web.Mvc → Microsoft.AspNetCore.Mvc
// 2. Manual service instantiation → Constructor injection with interfaces
// 3. Add ILogger for logging
// 4. All methods become async with Task<IActionResult>
// 5. ActionResult → IActionResult
// 6. HttpNotFound() → NotFound()
// 7. Use nameof() for action names in redirects
// 8. Add [FromForm] for form data binding (explicit)
// 9. Use DTOs instead of ViewModels for input
// 10. Nullable reference types enabled (string?)

File: packages/migration-cli/templates/skills/business-logic-mapping/examples/controller-example.java
// =============================================================================
// Java EE to Spring Boot Controller Conversion Example
// =============================================================================

// -----------------------------------------------------------------------------
// BEFORE: Java EE JAX-RS Resource (com/example/resources/ProductResource.java)
// -----------------------------------------------------------------------------
/*
package com.example.resources;

import com.example.models.Product;
import com.example.services.ProductService;

import javax.ejb.EJB;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;

@Path("/products")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class ProductResource {

    @EJB
    private ProductService productService;

    @GET
    public List<Product> getAllProducts() {
        return productService.getAllProducts();
    }

    @GET
    @Path("/{id}")
    public Response getProduct(@PathParam("id") Long id) {
        Product product = productService.getProductById(id);
        
        if (product == null) {
            return Response.status(Response.Status.NOT_FOUND).build();
        }
        
        return Response.ok(product).build();
    }

    @POST
    public Response createProduct(Product product) {
        Product created = productService.createProduct(product);
        return Response.status(Response.Status.CREATED)
                .entity(created)
                .build();
    }

    @PUT
    @Path("/{id}")
    public Response updateProduct(@PathParam("id") Long id, Product product) {
        Product updated = productService.updateProduct(id, product);
        
        if (updated == null) {
            return Response.status(Response.Status.NOT_FOUND).build();
        }
        
        return Response.ok(updated).build();
    }

    @DELETE
    @Path("/{id}")
    public Response deleteProduct(@PathParam("id") Long id) {
        boolean deleted = productService.deleteProduct(id);
        
        if (!deleted) {
            return Response.status(Response.Status.NOT_FOUND).build();
        }
        
        return Response.noContent().build();
    }

    @GET
    @Path("/search")
    public List<Product> searchProducts(
            @QueryParam("query") String query,
            @QueryParam("categoryId") Long categoryId,
            @QueryParam("minPrice") Double minPrice,
            @QueryParam("maxPrice") Double maxPrice) {
        
        return productService.searchProducts(query, categoryId, minPrice, maxPrice);
    }
}
*/

// -----------------------------------------------------------------------------
// AFTER: Spring Boot REST Controller (com/example/controller/ProductController.java)
// -----------------------------------------------------------------------------

package com.example.controller;

import com.example.dto.CreateProductDto;
import com.example.dto.UpdateProductDto;
import com.example.dto.ProductSearchFilters;
import com.example.model.Product;
import com.example.service.ProductService;
import jakarta.validation.Valid;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/products")
public class ProductController {

    private static final Logger log = LoggerFactory.getLogger(ProductController.class);
    
    private final ProductService productService;

    // Constructor injection (replaces @EJB)
    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    // GET /api/products
    @GetMapping
    public ResponseEntity<List<Product>> getAllProducts() {
        List<Product> products = productService.getAllProducts();
        return ResponseEntity.ok(products);
    }

    // GET /api/products/{id}
    @GetMapping("/{id}")
    public ResponseEntity<Product> getProduct(@PathVariable Long id) {
        return productService.getProductById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    // POST /api/products
    @PostMapping
    public ResponseEntity<Product> createProduct(@Valid @RequestBody CreateProductDto dto) {
        Product created = productService.createProduct(dto);
        log.info("Product created: {}", created.getId());
        return ResponseEntity.status(HttpStatus.CREATED).body(created);
    }

    // PUT /api/products/{id}
    @PutMapping("/{id}")
    public ResponseEntity<Product> updateProduct(
            @PathVariable Long id,
            @Valid @RequestBody UpdateProductDto dto) {
        
        return productService.updateProduct(id, dto)
                .map(product -> {
                    log.info("Product updated: {}", id);
                    return ResponseEntity.ok(product);
                })
                .orElse(ResponseEntity.notFound().build());
    }

    // DELETE /api/products/{id}
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteProduct(@PathVariable Long id) {
        if (productService.deleteProduct(id)) {
            log.info("Product deleted: {}", id);
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.notFound().build();
    }

    // GET /api/products/search
    @GetMapping("/search")
    public ResponseEntity<List<Product>> searchProducts(
            @RequestParam String query,
            @RequestParam(required = false) Long categoryId,
            @RequestParam(required = false) Double minPrice,
            @RequestParam(required = false) Double maxPrice) {
        
        ProductSearchFilters filters = new ProductSearchFilters(categoryId, minPrice, maxPrice);
        List<Product> results = productService.searchProducts(query, filters);
        return ResponseEntity.ok(results);
    }
}

// -----------------------------------------------------------------------------
// DTO for Create (dto/CreateProductDto.java)
// -----------------------------------------------------------------------------

package com.example.dto;

import jakarta.validation.constraints.*;
import java.math.BigDecimal;

public record CreateProductDto(
    @NotBlank
    @Size(max = 255)
    String name,

    @NotNull
    @DecimalMin("0.00")
    BigDecimal price,

    String description
) {}

// -----------------------------------------------------------------------------
// DTO for Update (dto/UpdateProductDto.java)
// -----------------------------------------------------------------------------

package com.example.dto;

import jakarta.validation.constraints.*;
import java.math.BigDecimal;

public record UpdateProductDto(
    @NotBlank
    @Size(max = 255)
    String name,

    @NotNull
    @DecimalMin("0.00")
    BigDecimal price,

    String description
) {}

// -----------------------------------------------------------------------------
// Search Filters DTO (dto/ProductSearchFilters.java)
// -----------------------------------------------------------------------------

package com.example.dto;

public record ProductSearchFilters(
    Long categoryId,
    Double minPrice,
    Double maxPrice
) {}

// -----------------------------------------------------------------------------
// KEY CONVERSION NOTES:
// -----------------------------------------------------------------------------
// 1. @Path → @RequestMapping / @GetMapping / @PostMapping etc.
// 2. @EJB → Constructor injection (Spring manages beans)
// 3. @PathParam → @PathVariable
// 4. @QueryParam → @RequestParam
// 5. Response.ok() → ResponseEntity.ok()
// 6. Response.status(NOT_FOUND) → ResponseEntity.notFound()
// 7. Response.status(CREATED) → ResponseEntity.status(HttpStatus.CREATED)
// 8. MediaType annotations → Spring handles JSON by default
// 9. Add @Valid for request body validation
// 10. Use Optional for nullable returns from service
// 11. Use Java records for DTOs (immutable, concise)
// 12. Add SLF4J logging (replaces custom logging)
